[quantization] Use more efficient specialized Quantize routine

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25751 [quantization] Factor unnecesary work out of add inner loop
* **#25731 [quantization] Use more efficient specialized Quantize routine**

I didn't notice this before, but the QuantizeAvx2 routine was requantizing only a single vector of 8 floats into 1/4 of a 256-bit int8 register. This switches it to use a specialization that goes from 4 float vectors into a whole int8 vector, borrowed from C2

Benchmark Script
```
import torch
import time

x = torch.rand(1, 256, 56, 56)
y = torch.rand(1, 256, 56, 56)

print('dtype', 'ms/iter (float)', 'ms/iter (quant)', 'quant / float', sep='\t')

for dtype in [torch.quint8, torch.qint8, torch.qint32]:
    qX = torch.quantize_linear(x, 0.1, 5, dtype).permute([0, 3, 1, 2])
    qY = torch.quantize_linear(y, 0.1, 5, dtype).permute([0, 3, 1, 2])

    _x = x.permute([0, 3, 1, 2])
    _y = y.permute([0, 3, 1, 2])

    NITER = 1000

    # Test float
    s = time.time()
    for i in range(NITER):
        _x + _y
    elapsed_float = time.time() - s
    ms_per_iter_float = elapsed_float / NITER * 1000    

    # Test quantized
    s = time.time()
    for i in range(NITER):
        torch.ops.quantized.add(qX, qY, 0.1, 5)
    elapsed = time.time() - s
    ms_per_iter = elapsed / NITER * 1000

    print(str(dtype), ms_per_iter_float, ms_per_iter, ms_per_iter / ms_per_iter_float, sep='\t')
    print('float gbps', (x.numel() + 2 * y.numel()) * x.element_size() / ms_per_iter_float / 1e6)
    print('quant gbps', (qX.numel() + 2 * qX.numel()) * qX.element_size() / ms_per_iter / 1e6)
```
Before this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.4862399101257324      0.34175705909729004     0.702856865469801
float gbps      quant gbps
19.812836830915185      7.0472516540305685
torch.qint8     0.46132922172546387     0.3334841728210449      0.7228767594078416
float gbps      quant gbps
20.882683225588192      7.222075877323351
torch.qint32    0.4637627601623535      4.034380197525024       8.699232763132326
float gbps      quant gbps
20.77310389611148       2.387923678068332
```
After this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.47435569763183594     0.20490765571594238     0.43197047434850966
float gbps      quant gbps
20.309215316893113      11.753821454766738
torch.qint8     0.4657437801361084      0.19967269897460938     0.4287179077652036
float gbps      quant gbps
20.684746444031163      12.061979491278683
torch.qint32    0.47335100173950195     4.043696641921997       8.54270219575315
float gbps      quant gbps
20.352321986426766      2.382422039310296
```

Differential Revision: [D17214413](https://our.internmc.facebook.com/intern/diff/D17214413)

Commit:
[quantization] Use more efficient specialized Quantize routine

Commit:
Update on "[quantization] Use more efficient specialized Quantize routine"

Commit:
Update on "[quantization] Use more efficient specialized Quantize routine"


I didn't notice this before, but the QuantizeAvx2 routine was requantizing only a single vector of 8 floats into 1/4 of a 256-bit int8 register. This switches it to use a specialization that goes from 4 float vectors into a whole int8 vector, borrowed from C2

Differential Revision: [D17214413](https://our.internmc.facebook.com/intern/diff/D17214413)

Commit:
Update on "[quantization] Use more efficient specialized Quantize routine"


I didn't notice this before, but the QuantizeAvx2 routine was requantizing only a single vector of 8 floats into 1/4 of a 256-bit int8 register. This switches it to use a specialization that goes from 4 float vectors into a whole int8 vector, borrowed from C2

Benchmark Script
```
import torch
import time

x = torch.rand(1, 256, 56, 56)
y = torch.rand(1, 256, 56, 56)

print('dtype', 'ms/iter (float)', 'ms/iter (quant)', 'quant / float', sep='\t')

for dtype in [torch.quint8, torch.qint8, torch.qint32]:
    qX = torch.quantize_linear(x, 0.1, 5, dtype).permute([0, 3, 1, 2])
    qY = torch.quantize_linear(y, 0.1, 5, dtype).permute([0, 3, 1, 2])

    _x = x.permute([0, 3, 1, 2])
    _y = y.permute([0, 3, 1, 2])

    NITER = 1000

    # Test float
    s = time.time()
    for i in range(NITER):
        _x + _y
    elapsed_float = time.time() - s
    ms_per_iter_float = elapsed_float / NITER * 1000    

    # Test quantized
    s = time.time()
    for i in range(NITER):
        torch.ops.quantized.add(qX, qY, 0.1, 5)
    elapsed = time.time() - s
    ms_per_iter = elapsed / NITER * 1000

    print(str(dtype), ms_per_iter_float, ms_per_iter, ms_per_iter / ms_per_iter_float, sep='\t')
    print('float gbps', (x.numel() + 2 * y.numel()) * x.element_size() / ms_per_iter_float / 1e6)
    print('quant gbps', (qX.numel() + 2 * qX.numel()) * qX.element_size() / ms_per_iter / 1e6)
```
Before this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.4862399101257324      0.34175705909729004     0.702856865469801
float gbps      quant gbps
19.812836830915185      7.0472516540305685
torch.qint8     0.46132922172546387     0.3334841728210449      0.7228767594078416
float gbps      quant gbps
20.882683225588192      7.222075877323351
torch.qint32    0.4637627601623535      4.034380197525024       8.699232763132326
float gbps      quant gbps
20.77310389611148       2.387923678068332
```
After this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.47435569763183594     0.20490765571594238     0.43197047434850966
float gbps      quant gbps
20.309215316893113      11.753821454766738
torch.qint8     0.4657437801361084      0.19967269897460938     0.4287179077652036
float gbps      quant gbps
20.684746444031163      12.061979491278683
torch.qint32    0.47335100173950195     4.043696641921997       8.54270219575315
float gbps      quant gbps
20.352321986426766      2.382422039310296
```

Differential Revision: [D17214413](https://our.internmc.facebook.com/intern/diff/D17214413)

llyfacebook commented on aten/src/ATen/cpu/vec256/vec256_qint.h:
```diff
@@ -39,73 +39,92 @@ namespace {
+    T max_val);
+
+template <>
+__m256i pack_saturate_and_clamp<int8_t>(
```
Can we also inline this function? Maybe compiler can decide, :)

lly-zero-one reviewed (commented):
Great. You switch the order of clip and round, which seems to be equivalent. Maybe double check the saturated value. If it is better than the original implementation, can we also suggest fbgemm quantize to use the same way?

Commit:
Update on "[quantization] Use more efficient specialized Quantize routine"


I didn't notice this before, but the QuantizeAvx2 routine was requantizing only a single vector of 8 floats into 1/4 of a 256-bit int8 register. This switches it to use a specialization that goes from 4 float vectors into a whole int8 vector, borrowed from C2

Benchmark Script
```
import torch
import time

x = torch.rand(1, 256, 56, 56)
y = torch.rand(1, 256, 56, 56)

print('dtype', 'ms/iter (float)', 'ms/iter (quant)', 'quant / float', sep='\t')

for dtype in [torch.quint8, torch.qint8, torch.qint32]:
    qX = torch.quantize_linear(x, 0.1, 5, dtype).permute([0, 3, 1, 2])
    qY = torch.quantize_linear(y, 0.1, 5, dtype).permute([0, 3, 1, 2])

    _x = x.permute([0, 3, 1, 2])
    _y = y.permute([0, 3, 1, 2])

    NITER = 1000

    # Test float
    s = time.time()
    for i in range(NITER):
        _x + _y
    elapsed_float = time.time() - s
    ms_per_iter_float = elapsed_float / NITER * 1000    

    # Test quantized
    s = time.time()
    for i in range(NITER):
        torch.ops.quantized.add(qX, qY, 0.1, 5)
    elapsed = time.time() - s
    ms_per_iter = elapsed / NITER * 1000

    print(str(dtype), ms_per_iter_float, ms_per_iter, ms_per_iter / ms_per_iter_float, sep='\t')
    print('float gbps', (x.numel() + 2 * y.numel()) * x.element_size() / ms_per_iter_float / 1e6)
    print('quant gbps', (qX.numel() + 2 * qX.numel()) * qX.element_size() / ms_per_iter / 1e6)
```
Before this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.4862399101257324      0.34175705909729004     0.702856865469801
float gbps      quant gbps
19.812836830915185      7.0472516540305685
torch.qint8     0.46132922172546387     0.3334841728210449      0.7228767594078416
float gbps      quant gbps
20.882683225588192      7.222075877323351
torch.qint32    0.4637627601623535      4.034380197525024       8.699232763132326
float gbps      quant gbps
20.77310389611148       2.387923678068332
```
After this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.47435569763183594     0.20490765571594238     0.43197047434850966
float gbps      quant gbps
20.309215316893113      11.753821454766738
torch.qint8     0.4657437801361084      0.19967269897460938     0.4287179077652036
float gbps      quant gbps
20.684746444031163      12.061979491278683
torch.qint32    0.47335100173950195     4.043696641921997       8.54270219575315
float gbps      quant gbps
20.352321986426766      2.382422039310296
```

Differential Revision: [D17214413](https://our.internmc.facebook.com/intern/diff/D17214413)

Commit:
Update on "[quantization] Use more efficient specialized Quantize routine"


I didn't notice this before, but the QuantizeAvx2 routine was requantizing only a single vector of 8 floats into 1/4 of a 256-bit int8 register. This switches it to use a specialization that goes from 4 float vectors into a whole int8 vector, borrowed from C2

Benchmark Script
```
import torch
import time

x = torch.rand(1, 256, 56, 56)
y = torch.rand(1, 256, 56, 56)

print('dtype', 'ms/iter (float)', 'ms/iter (quant)', 'quant / float', sep='\t')

for dtype in [torch.quint8, torch.qint8, torch.qint32]:
    qX = torch.quantize_linear(x, 0.1, 5, dtype).permute([0, 3, 1, 2])
    qY = torch.quantize_linear(y, 0.1, 5, dtype).permute([0, 3, 1, 2])

    _x = x.permute([0, 3, 1, 2])
    _y = y.permute([0, 3, 1, 2])

    NITER = 1000

    # Test float
    s = time.time()
    for i in range(NITER):
        _x + _y
    elapsed_float = time.time() - s
    ms_per_iter_float = elapsed_float / NITER * 1000    

    # Test quantized
    s = time.time()
    for i in range(NITER):
        torch.ops.quantized.add(qX, qY, 0.1, 5)
    elapsed = time.time() - s
    ms_per_iter = elapsed / NITER * 1000

    print(str(dtype), ms_per_iter_float, ms_per_iter, ms_per_iter / ms_per_iter_float, sep='\t')
    print('float gbps', (x.numel() + 2 * y.numel()) * x.element_size() / ms_per_iter_float / 1e6)
    print('quant gbps', (qX.numel() + 2 * qX.numel()) * qX.element_size() / ms_per_iter / 1e6)
```
Before this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.4862399101257324      0.34175705909729004     0.702856865469801
float gbps      quant gbps
19.812836830915185      7.0472516540305685
torch.qint8     0.46132922172546387     0.3334841728210449      0.7228767594078416
float gbps      quant gbps
20.882683225588192      7.222075877323351
torch.qint32    0.4637627601623535      4.034380197525024       8.699232763132326
float gbps      quant gbps
20.77310389611148       2.387923678068332
```
After this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.47435569763183594     0.20490765571594238     0.43197047434850966
float gbps      quant gbps
20.309215316893113      11.753821454766738
torch.qint8     0.4657437801361084      0.19967269897460938     0.4287179077652036
float gbps      quant gbps
20.684746444031163      12.061979491278683
torch.qint32    0.47335100173950195     4.043696641921997       8.54270219575315
float gbps      quant gbps
20.352321986426766      2.382422039310296
```

Differential Revision: [D17214413](https://our.internmc.facebook.com/intern/diff/D17214413)

Commit:
Update on "[quantization] Use more efficient specialized Quantize routine"


I didn't notice this before, but the QuantizeAvx2 routine was requantizing only a single vector of 8 floats into 1/4 of a 256-bit int8 register. This switches it to use a specialization that goes from 4 float vectors into a whole int8 vector, borrowed from C2

Benchmark Script
```
import torch
import time

x = torch.rand(1, 256, 56, 56)
y = torch.rand(1, 256, 56, 56)

print('dtype', 'ms/iter (float)', 'ms/iter (quant)', 'quant / float', sep='\t')

for dtype in [torch.quint8, torch.qint8, torch.qint32]:
    qX = torch.quantize_linear(x, 0.1, 5, dtype).permute([0, 3, 1, 2])
    qY = torch.quantize_linear(y, 0.1, 5, dtype).permute([0, 3, 1, 2])

    _x = x.permute([0, 3, 1, 2])
    _y = y.permute([0, 3, 1, 2])

    NITER = 1000

    # Test float
    s = time.time()
    for i in range(NITER):
        _x + _y
    elapsed_float = time.time() - s
    ms_per_iter_float = elapsed_float / NITER * 1000    

    # Test quantized
    s = time.time()
    for i in range(NITER):
        torch.ops.quantized.add(qX, qY, 0.1, 5)
    elapsed = time.time() - s
    ms_per_iter = elapsed / NITER * 1000

    print(str(dtype), ms_per_iter_float, ms_per_iter, ms_per_iter / ms_per_iter_float, sep='\t')
    print('float gbps', (x.numel() + 2 * y.numel()) * x.element_size() / ms_per_iter_float / 1e6)
    print('quant gbps', (qX.numel() + 2 * qX.numel()) * qX.element_size() / ms_per_iter / 1e6)
```
Before this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.4862399101257324      0.34175705909729004     0.702856865469801
float gbps      quant gbps
19.812836830915185      7.0472516540305685
torch.qint8     0.46132922172546387     0.3334841728210449      0.7228767594078416
float gbps      quant gbps
20.882683225588192      7.222075877323351
torch.qint32    0.4637627601623535      4.034380197525024       8.699232763132326
float gbps      quant gbps
20.77310389611148       2.387923678068332
```
After this change (DynDisp to AVX2)
```
dtype   ms/iter (float) ms/iter (quant) quant / float
torch.quint8    0.47435569763183594     0.20490765571594238     0.43197047434850966
float gbps      quant gbps
20.309215316893113      11.753821454766738
torch.qint8     0.4657437801361084      0.19967269897460938     0.4287179077652036
float gbps      quant gbps
20.684746444031163      12.061979491278683
torch.qint32    0.47335100173950195     4.043696641921997       8.54270219575315
float gbps      quant gbps
20.352321986426766      2.382422039310296
```

Differential Revision: [D17214413](https://our.internmc.facebook.com/intern/diff/D17214413)
